docs/source/core/adapters/components.rst

```diff
@@ -0,0 +1,90 @@
+
+1) Functional form - the trainable parameters that will modify the input
+2) Insertion form - Where the adapter outputs are integrated with the original input. The input to the adapters can be the last output of the layer, the input to some attention layer, or even the original input to the module itself (before even the modules forward pass).
+3) Composition function - How the adapters outputs are integrated with the inputs. It can be as simple as residual addition connection, or concatenation, or point-wise multiplication etc.
+
+Functional Form - Adapter Networks
+==================================
+
+Adapter modules represent the functional form of the adapter. We discuss an example of a most commonly used adapter module found in literature, titled the ``LinearAdapter`` (or Houlsby Adapter) :cite:`adapters-houlsby2019adapter`.
+
+.. note::
+
+    All adapter modules must extend :class:`~nemo.collections.common.parts.adapter_modules.AbstractAdapterModule` and should ideally have an equivalent DataClass config for easy instantiation !
+
+
+.. autoclass:: nemo.collections.common.parts.adapter_modules.AbstractAdapterModule
+    :show-inheritance:
+    :members:
+    :member-order: bysource
+
+-----
+
+.. autoclass:: nemo.collections.common.parts.adapter_modules.LinearAdapter
+    :show-inheritance:
+    :members:
+    :member-order: bysource
+
+
+Insertion Form - Module Adapters
+--------------------------------
+
+Adapter modules can be integrated into many different locations of a given module. For example, it is possible to have an adapter that affects only the outputs of the final layer in each module. We can also have a ``Parallel Adapter`` :cite:`adapters-Junxian2021unified` that operates at the input of the module itself, in parallel to the forward pass of the module. Yet another insertion location is inside the Multi Head Attention Layers.
+
+On top of this, while adapters are commonly used only in the layers containing the most parameters (say the Encoder of a network), some models can support adapters in multiple locations (Encoder-Decoder architecture for Language Models, Machine Translation, or even Encoder-Decoder-Joint for ASR with Transducer Loss). As such, NeMo utilizes the concept of ``Module Adapters``.
+
+``Module Adapters`` are very simply defined when adding an adapter - by specifying the module that the adapter should be inserted into.
+
+.. code-block:: python
+
+    # Get the list of supported modules / locations in a adapter compatible Model
+    print(model.adapter_module_names)  # assume ['', 'encoder', 'decoder']
+
+    # When calling add_adapter, specify the module name in the left of the colon symbol, and the adapter name afterwords.
+    # The adapter is then directed to the decoder module instead of the default / encoder module.
+    model.add_adapter("decoder:first_adapter", cfg=...)
```

Comment:
consider adding similar code snippets in functional and insertion forms.

Reply:
Hmm those are internal details, users will almost never see them during usage, not even when creating their own adapter modules